### ISSUE TITLE
Request::getIp() trusts client headers and PSR-7 helper methods

### ISSUE THREAD

**Issue by AndrewCarterUK** [slimphp/Slim#1528]
https://github.com/slimphp/Slim/blob/3.x/Slim/Http/Request.php#L1078

It would be trivial for a client to send an X-Forwarded-For header that tricked the application into thinking it was from somewhere else. If this is going to be done, a trusted proxies whitelist must be established - and the method should check to see that the request comes from one of these before trusting the contents of the header.

My personal opinion is that adding methods like this to the PSR-7 implementation makes no sense. The whole point of using PSR-7 is that you are writing code which isn't coupled to an implementation of the standard. I think it makes more sense to create a seperate helper class which takes the PSR-7 message as an input.

**Comment by gkralik** [comment]
Maybe some method on the Request to set trusted proxy addresses?
Additionally, the `X-Forwarded-For` header can contain multiple addresses, eg:

```
X-Forwarded-For: client, firstProxy, secondProxy
```

So we should probably check for that too and only return the client IP address if the proxy addresses are trusted.

**Comment by AndrewCarterUK** [comment]
You wouldn't want to store trusted proxy addresses inside the request message, as they are not part of the request. Maybe something more like:

``` php
public function getIp(array $trustedProxies = []);
```

I'm personally against the helpers in their current format all together, I think it would be better to move these into a separate class.

The way the Slim request is explained in the documentation (http://www.slimframework.com/docs/objects/request.html) is also a bit misleading.

The page starts with an example of how to get the request object in middleware, like so:

``` php
<?php
use Psr\Http\Message\ServerRequestInterface;
use Psr\Http\Message\ResponseInterface;

$app = new \Slim\App;
$app->add(function (ServerRequestInterface $request, ResponseInterface $response, callable $next) {
    // Use the PSR 7 $request object

    return $next($request, $response);
});
// Define app routes...
$app->run();
```

And then goes on to give code examples like this:

``` php
if ($request->isXhr()) {
    // Do something
}

$contentType = $request->getContentType();

// etc...
```

I think that these helper methods should be defined in separate interfaces that extend the PSR-7 interfaces.

```
- Psr\Http\Message\ServerRequestInterface
  - Slim\Interfaces\Http\ServerRequestInterface
- Psr\Http\Message\ResponseInterface
  - Slim\Interfaces\Http\ResponseInterface
```

The documentation should encourage developers using the helper functions to type hint using the Slim interfaces and not the PSR-7 ones.

Ultimately though I think it would be better to build a PSR-7 helper component - then it could be used with PSR-7 messages from other libraries.

**Comment by silentworks** [comment]
You are allowed to add non PSR-7 methods to the implementation, the framework itself is not dependent on any of these methods as I have swapped out our implementation for diactoros and it all worked as expected. Zend Expressive also has its own helper methods for their PSR-7 implementation although a bit more separated than we have.

**Comment by AndrewCarterUK** [comment]
@silentworks being allowed to add non PSR-7 methods to the implementation doesn't make it a good idea to do so. If it is done, these methods should definitely be defined in an interface that applications can type hint against.

Zend Expressive doesn't provide a PSR-7 implementation, it uses Zend Diactoros.

As an example of why this is a bad idea,  I've been working on a library that allows developers to write FastCGI applications. It has a Slim v3 adapter (https://github.com/PHPFastCGI/Slimmer) that injects PSR-7 requests into the application. Any code using these helper methods would cause an issue as the requests are created with Zend Diactoros (edit: even if the framework runs, applications using the methods would not). Using a wrapper class would solve that problem.

Also, the middleware stack does not guarantee that as the PSR-7 request message is passed along the pipe it will retain the original class type. It is absolutely possible for a piece of middleware to completely replace the request message with one generated by a different PSR-7 implementation. Again, this is solved if a wrapper class is used to provide the helpers.

[edit] Looking at Zend Diactoros it doesn't appear to provide helper methods. If you read this issue - https://github.com/zendframework/zend-diactoros/issues/44 - the core contributor (and the editor of PSR-7) says "I feel such stuff should be in add-on packages; psr-7 was written in part to promote such single purpose utilities and prevent writing framework-specific code."

**Comment by akrabat** [comment]
It's actually Stratigility that provides the wrappers that @silentworks is talking about:

https://github.com/zendframework/zend-stratigility/blob/master/src/Http/Request.php

**Comment by akrabat** [comment]
Pinging @codeguy for his thoughts on this.

**Comment by akrabat** [comment]
There's two issues here.
1. Whether Slim\Http should extend PSR-7 without its own Interface.
2. Potential security issues in `getIp()` due to looking for `X-Forwarded-For`

Item 1 can be targeted at a later 3.x release as it won't break BC.

For item 2, my view is that we should remove `getIp()` altogether and then someone could write middleware that sets an "ip" attribute into the `Request` object if they wanted to.

Alternatively, I would support removing  the `X-Forwarded-For` test in `getIp()` and adding a `setIp()` method, so, again, someone could write middleware that can set the ip to something other than `REMOTE_ADDR`.

**Comment by AndrewCarterUK** [comment]
@akrabat

With regards to number 1, what you say is only true if you accept that Slim should add methods to PSR-7 **at all**, which I think that it almost definitely shouldn't. If developers are going to be encouraged to build against a Slim implementation of PSR-7 that goes beyond the interfaces, they are not bulding against PSR-7 at all, they are building against a custom Slim HTTP library - which completely defeats the purpose.

Also, as soon as these methods are included in a 3.x release, they cannot be removed without breaking B/C. If there is doubt over their existence, it would make more sense to start with them removed and then add them in.

Number 2, I agree it should be removed. The middleware approach that you have suggested is a good idea - could this work with the other helper methods too?

I'm not sure I agree with the second approach. To start with, it would be `withIp()`, and it's also another method that would prevent developers from writing portable code.

I'm happy to work on a pull request for a helper class or helper middleware that sets attributes on the request if this would be useful?

**Comment by akrabat** [comment]
@AndrewCarterUK I'd love so see the middleware. It would be a separate repository from Slim 3 as it's generically useful for any PSR-7 based middleware stack.

**Comment by AndrewCarterUK** [comment]
I'll create a new repository using the Slim-Csrf repo as a guide, then you can decide if you want to transfer the repository into the Slim namespace or not.

I'm also going to include helper classes that would allow the Slim PSR-7 implementations of the Request and Response object to be stripped back to the interfaces, e.g.

``` php
if ($request->isPost()) {
    // ...
}
```

Becomes

``` php
if ($requestHelper->isPost($request)) {
    // ...
}
```

At least then can decide whether to include or strip them.

**Comment by akrabat** [comment]
My take on middleware that is a better implementation of `getIp()`: https://github.com/akrabat/rka-ip-address-middleware

**Comment by akrabat** [comment]
https://github.com/oscarotero/psr7-middlewares/blob/master/src/Middleware/ClientIp.php is another one.

**Comment by gkralik** [comment]
:+1:

**Comment by r3wt** [comment]
You should probably add attribution to the person who originally authored this code(stack overflow), you can optimize this by inlining the filter function as well

Sent from my Windows Phone

**Comment by akrabat** [comment]
@r3wt: Given that I haven't looked at Stack Overflow, can you point me at the page you think that I've copied?

**Comment by r3wt** [comment]
I recognize the function from a famous answer on stack. I have identical code in many projects, alot of people are using this function. I can't link from mobile but you should be able to find I by searching get up on stack

Sent from my Windows Phone

**Comment by akrabat** [comment]
I'm confused. You have https://github.com/akrabat/rka-ip-address-middleware/blob/master/src/IpAddress.php#L75-L105 in your projects?

**Comment by r3wt** [comment]
I'll link you she I get home
Sent from my Windows Phone

**Comment by r3wt** [comment]
here's the function in question:

``` php

function getIP()
{
    foreach (array(
        'HTTP_CLIENT_IP',
        'HTTP_X_FORWARDED_FOR',
        'HTTP_X_FORWARDED',
        'HTTP_X_CLUSTER_CLIENT_IP',
        'HTTP_FORWARDED_FOR',
        'HTTP_FORWARDED',
        'REMOTE_ADDR'
    ) as $key) {
        if (array_key_exists($key, $_SERVER) === true) {
            foreach (array_map('trim', explode(',', $_SERVER[$key])) as $ip) {
                if (filter_var($ip, FILTER_VALIDATE_IP, FILTER_FLAG_NO_PRIV_RANGE | FILTER_FLAG_NO_RES_RANGE) !== false) {
                    return $ip;
                }
            }
        }
    }
}
```

can't find the original answer on stack overflow. Fyi, i wasn't insinuating anything, just thought that if you copied the same answer i did it would be fair to add attribution since this is open source.

check similarity to https://github.com/akrabat/rka-ip-address-middleware/blob/master/src/IpAddress.php#L91-101

**Comment by akrabat** [comment]
Presumably there's not that many ways to actually do this :)

**Comment by akrabat** [comment]
Fixed in #1559
